### PR TITLE
MSF-23675: LCM Bricks synchronizes groups and permissions of dashboards from the source workspace to target workspace

### PR DESCRIPTION
**Pull request by dangdinh**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1137/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1018/ : SUCCESS in 13m 47s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1139/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1020/ : SUCCESS in 12m 19s

File: lib/gooddata/lcm/actions/synchronize_user_groups.rb [hunk 1]
```diff
@@ -27,31 +27,44 @@ class SynchronizeUserGroups < BaseAction
 
         description 'Additional Hidden Parameters'
         param :additional_hidden_params, instance_of(Type::HashType), required: false
+
+        description 'Disable dashboard permission for pixel perfect dashboard'
```
Comment (danh-ung) [inline lib/gooddata/lcm/actions/synchronize_user_groups.rb:31]: Disable synchronizing dashboard permissions for Pixel Perfect dashboards.
Disable synchronizing dashboard permissions for AD/KD dashboards.

File: lib/gooddata/lcm/actions/synchronize_user_groups.rb [hunk 2]
```diff
@@ -27,31 +27,44 @@ class SynchronizeUserGroups < BaseAction
...
       class << self
         def call(params)
           results = ThreadSafe::Array.new
 
-          client = params.gdc_gd_client
-          development_client = params.development_client
+          disable_pp_dashboard_permission = GoodData::Helpers.to_boolean(params.disable_pp_dashboard_permission)
+          disable_kd_dashboard_permission = GoodData::Helpers.to_boolean(params.disable_kd_dashboard_permission)
+
+          if disable_pp_dashboard_permission && disable_kd_dashboard_permission
```
Comment (danh-ung) [inline lib/gooddata/lcm/actions/synchronize_user_groups.rb:45]: User groups is always synchronized before, we shouldn't change this behaviour, no need to check params disable_pp_dashboard_permission && disable_kd_dashboard_permission

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1141/ : SUCCESS in 14s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1021/ : FAILURE in 10m 49s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1143/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1023/ : FAILURE in 10m 48s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1155/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1035/ : SUCCESS in 12m 42s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1158/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1038/ : SUCCESS in 13m 31s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1160/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1040/ : SUCCESS in 11m 03s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1172/ : SUCCESS in 13s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1051/ : FAILURE in 11m 32s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1174/ : SUCCESS in 13s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1053/ : SUCCESS in 11m 05s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1178/ : SUCCESS in 9s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1056/ : SUCCESS in 12m 59s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1180/ : SUCCESS in 13s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1058/ : FAILURE in 11m 05s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1182/ : SUCCESS in 13s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1060/ : SUCCESS in 11m 13s

File: lib/gooddata/models/project.rb [hunk 3]
```diff
@@ -436,6 +426,68 @@ def transfer_user_groups(from_project, to_project)
...
+
+            revoke_group_dashboard_permission(target_group_dashboards, target_dashboard, common_group_names)
+            add_group_dashboard_permission(source_group_dashboards, target_dashboard, common_group_names, target_user_groups)
+          rescue StandardError => e
+            GoodData.logger.warn "Failed to synchronize #{dashboard_type} dashboard permission from project: '#{from_project.title}', PID: '#{from_project.pid}' to project: '#{to_project.title}', PID: '#{to_project.pid}', dashboard: '#{target_dashboard.title}', dashboard uri: '#{target_dashboard.uri}' . Error: #{e.message} - #{e}" # rubocop:disable Metrics/LineLength
+          end
+        end
+      end
+
+      def user_group_dashboards(user_groups, dashboard)
```
Comment (sangtm) [inline lib/gooddata/models/project.rb:451]: --> get_dashboard_user_groups

File: lib/gooddata/models/project.rb [hunk 4]
```diff
@@ -436,6 +426,68 @@ def transfer_user_groups(from_project, to_project)
         end
       end
 
+      def transfer_dashboard_permission(from_project, to_project, source_dashboards, target_dashboards, dashboard_type)
```
Comment (sangtm) [inline lib/gooddata/models/project.rb:429]: remove dashboard_type param

File: lib/gooddata/models/project.rb [hunk 5]
```diff
@@ -436,6 +426,68 @@ def transfer_user_groups(from_project, to_project)
...
+          group_dashboards << {
+            name: user_group.name,
+            user_group: user_group,
+            permission: permission
+          }
+        end
+        group_dashboards
+      end
+
+      def revoke_group_dashboard_permission(group_dashboards, dashboard, common_group_names)
```
Comment (sangtm) [inline lib/gooddata/models/project.rb:470]: --> remove_user_groups_from_dashboard

File: lib/gooddata/models/project.rb [hunk 6]
```diff
@@ -436,6 +426,68 @@ def transfer_user_groups(from_project, to_project)
...
+        group_dashboards.each do |group_dashboard|
+          group_name = group_dashboard[:name]
+          next if common_group_names && common_group_names[group_name]
+
+          dashboard.revoke(:member => group_dashboard[:user_group], :permission => group_dashboard[:permission])
+        end
+      end
+
+      def add_group_dashboard_permission(group_dashboards, dashboard, common_group_names, target_user_groups)
```
Comment (sangtm) [inline lib/gooddata/models/project.rb:479]: add_user_groups_to_dashboard

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1186/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1064/ : FAILURE in 11m 16s

**Comment by dangdinh** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1187/ : SUCCESS in 6s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1065/ : SUCCESS in 11m 10s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1189/ : SUCCESS in 7s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1067/ : SUCCESS in 15m 38s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1191/ : SUCCESS in 10s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1069/ : SUCCESS in 13m 07s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-oss-check/1193/ : SUCCESS in 8s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-build-pipeline/1071/ : SUCCESS in 11m 18s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ruby-gate-PR1870

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-lcm-promote-pipeline/180/ : SUCCESS in 14m 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-sonarqube-scan-zuul-gate/95/ : SUCCESS in 5m 24s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-merge-compliance-check/637/ : SUCCESS in 23s
- https://jenkins-ii.intgdc.com/job/gooddata-ruby/job/gooddata-ruby-gitops-deploy-pipeline/60/ : SUCCESS in 2m 09s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-fossa-scanning-zuul-el7/180/ : FAILURE in 4m 09s (non-voting)

**Pull request merged**